### PR TITLE
FIX eagerLoad crash with nonterminal hasOne relation

### PR DESCRIPTION
**Pull request by MasonD**
Addresses #11470

### CODE REVIEW

**Comment by MasonD** [comment]
Looks like the failing tests are because the number of queries is no longer as previously expected. I'd appreciate input on how what I've done here has caused extra queries.

File: src/ORM/DataList.php [hunk 1]
```diff
@@ -1055,7 +1055,14 @@ private function fetchEagerLoadRelations(Query $query): void
             $parentIDs = $topLevelIDs;
             $parentRelationData = $query;
             $chainToDate = [];
+            $polymorphicEncountered = false;
             foreach (explode('.', $relationChain) as $relationName) {
+                if ($polymorphicEncountered) {
+                    $polymorphicRelation = $chainToDate[count($chainToDate) - 1];
```
Comment (GuySartorelli) [inline src/ORM/DataList.php]:
```suggestion
                    $polymorphicRelation = $chainToDate[array_key_last($chainToDate)];
```
This is a stylistic suggestion so I won't block merging if it's not done, but it is more easily readable at a glance.

**Comment by GuySartorelli** [comment]
> Looks like the failing tests are because the number of queries is no longer as previously expected. I'd appreciate input on how what I've done here has caused extra queries.

My guess, having looked briefly at the code changes but not at the tests, is that the tests were actually _not_ correct, and chains with `has_one` weren't actually checking anything after the `has_one` portion.
I don't have time to dive into the tests right now to validate that - can you please try to check whether the new results are correct or not?

**Comment by MasonD** [comment]
After investigation, the tests indeed did their job and caught a bug in my code, now fixed.

**Comment by MasonD** [comment]
@GuySartorelli I did resolve the failing tests, in case you missed it. The tests were correct and I was missing out on some eager fetching because of a bug.

**Comment by GuySartorelli** [comment]
Thanks yup. I saw that just haven't had a chance to come back to this yet.

**Review by GuySartorelli — CHANGES_REQUESTED** [review]
Looks good and seems to work as expected, thanks for taking the time to do this.

Couple of things:

1. Can you please add some unit tests? One to check that this works as expected for `HasOne.HasMany` (i.e. the intended use case) and one to check that the exception is thrown for `PolymorphicHasOne.HasMany`?
2. Can you please retarget this to the `5.3` branch? That way the patch can be released immediately rather than waiting for the April minor release.
You may need to [reset your commits](https://gist.github.com/GuySartorelli/74627843aeb75e12b777b1f44cb22d2f) after retargeting the PR.

**Comment by MasonD** [comment]
@GuySartorelli all done!

**Review by GuySartorelli — APPROVED** [review]
LGTM. Note the changes in the "expected" values are a result of adding new records to the `createEagerLoadData()` method which therefore had to be traversed. No additional queries were added to the eagerloading logic itself.

**Pull request merged**